vercel/resource_project_environment_variable.go

```diff
@@ -0,0 +1,239 @@
+			"Provider not configured",
+			"The provider hasn't been configured before apply. This leads to weird stuff happening, so we'd prefer if you didn't do that. Thanks!",
+		)
+		return
+	}
+
+	var plan ProjectEnvironmentVariable
+	diags := req.Plan.Get(ctx, &plan)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	response, err := r.p.client.UpsertEnvironmentVariable(ctx, plan.ProjectID.Value, plan.TeamID.Value, plan.toUpsertEnvironmentVariableRequest())
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Error creating project environment variable",
+			"Could not create project environment variable, unexpected error: "+err.Error(),
+		)
+		return
+	}
+
+	result := convertResponseToProjectEnvironmentVariable(response, plan)
+
+	tflog.Trace(ctx, "created project environment variable", map[string]interface{}{
+		"id":         result.ID.Value,
+		"team_id":    result.TeamID.Value,
+		"project_id": result.ProjectID.Value,
+	})
+
+	diags = resp.State.Set(ctx, result)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+}
+
+// Read will read an environment variable of a Vercel project by requesting it from the Vercel API, and will update terraform
+// with this information.
+func (r resourceProjectEnvironmentVariable) Read(ctx context.Context, req resource.ReadRequest, resp *resource.ReadResponse) {
+	var state ProjectEnvironmentVariable
+	diags := req.State.Get(ctx, &state)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	out, err := r.p.client.GetEnvironmentVariable(ctx, state.ProjectID.Value, state.TeamID.Value, state.ID.Value)
+
+	if client.NotFound(err) {
+		resp.State.RemoveResource(ctx)
+		return
+	}
+
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Error reading project environment variable",
+			fmt.Sprintf("Could not get project environment variable %s %s %s, unexpected error: %s",
+				state.ID.Value,
+				state.ProjectID.Value,
+				state.TeamID.Value,
+				err,
+			),
+		)
+		return
+	}
+
+	result := convertResponseToProjectEnvironmentVariable(out, state)
+	tflog.Trace(ctx, "read project environment variable", map[string]interface{}{
+		"id":         result.ID.Value,
+		"team_id":    result.TeamID.Value,
+		"project_id": result.ProjectID.Value,
+	})
+
+	diags = resp.State.Set(ctx, result)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+}
+
+// Update updates the environment variable of a Vercel project state.
+func (r resourceProjectEnvironmentVariable) Update(ctx context.Context, req resource.UpdateRequest, resp *resource.UpdateResponse) {
+	var plan ProjectEnvironmentVariable
+	diags := req.Plan.Get(ctx, &plan)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	var state ProjectEnvironmentVariable
+	diags = req.State.Get(ctx, &state)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
```

Comment:
You don't use `state` after reading it (this is only necessary if we want to work out what has changed). Since we just upsert I think you only care about the `plan` here.